### PR TITLE
Feature/redis keys analyzer

### PR DESCRIPTION
**Pull request by adrian-larysz-rmtcfm-com**

### CODE REVIEW

File: server/src/services/event/EventService.js [hunk 1]
```diff
@@ -1,27 +1,62 @@
...
         this.redisService = redisService;
+        this.actionService = new ActionService(prefix, redisService);
     }
 
     get stores() {
         const redisService = this.redisService;
+        const actionService = this.actionService;
 
         return [
             function (prefix, key, userId) {
```
Comment (rafal-ksiazek-rmtcfm-com) [inline server/src/services/event/EventService.js:19]: you can check arrow function, then "this" should be assigned to class automatically, but yeah, it's strange

Comment (rafal-ksiazek-rmtcfm-com) [reply server/src/services/event/EventService.js:19]: eventually you can always do:
const that = this;

Comment (rafal-ksiazek-rmtcfm-com) [reply server/src/services/event/EventService.js:19]: but of course, as you have it right now is also good

File: server/src/services/event/EventService.js [hunk 2]
```diff
@@ -1,27 +1,62 @@
...
 const timeSpans = require('./timeSpans');
 const scopes = require('./scopes');
+const AnalyzerService = require('./AnalyzerService');
+const ActionService = require('./ActionService');
 
 class EventService {
     constructor(prefix, redisService) {
         this.prefix = prefix;
         this.redisService = redisService;
+        this.actionService = new ActionService(prefix, redisService);
```
Comment (rafal-ksiazek-rmtcfm-com) [inline server/src/services/event/EventService.js:11]: violation of OCP

File: server/src/services/event/AnalyzerService.js [hunk 3]
```diff
@@ -7,6 +7,36 @@ class AnalyzerService {
...
+            list: key => {
+                return redisService.getListValues(key);
+            },
+
+            increment: key => {
+                return redisService.get(key).then(value => (value ? parseInt(value) : 0));
+            },
+
+            bitmap: key => {
+                return redisService.count(key);
```
Comment (rafal-ksiazek-rmtcfm-com) [inline server/src/services/event/AnalyzerService.js:31]: redisService.countBit(key)

File: server/src/services/event/ActionService.js [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        return this.redisService.getSortedSetLenght(`${this.prefix}:_private:userActions:${userId}:${action}`, {
+            min,
+            max
+        });
+    }
+
+    _calculateDateScore(date) {
+        const _date = dayjs(date);
+
+        return _date.year() * 2 + _date.month() * 30 + _date.date();
```
Comment (rafal-ksiazek-rmtcfm-com) [inline server/src/services/event/ActionService.js:44]: what's this?

File: server/src/services/event/ActionService.js [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+            min,
+            max
+        });
+    }
+
+    countAfter(userId, action, date) {
+        const min = `(${this._calculateDateScore(date)}`;
+        const max = '+inf';
+
+        return this.redisService.getSortedSetLenght(`${this.prefix}:_private:userActions:${userId}:${action}`, {
```
Comment (rafal-ksiazek-rmtcfm-com) [inline server/src/services/event/ActionService.js:35]: length

File: server/src/services/event/EventService.js [hunk 6]
```diff
@@ -32,20 +67,26 @@ class EventService {
...
                 const scopeName = scope !== scopesKey ? `:${scope}` : '';
+                const timeSpanName = _timeSpan !== timeSpansKey ? `:${_timeSpan}` : '';
 
-                return `${scopesKey}${scopeName}:${timeSpansKey}:${timeSpan(dayjs(date))}`;
+                return `${scopesKey}${scopeName}:${timeSpansKey}${timeSpanName}`;
             });
         });
 
+        if (args.action) {
+            await this.actionService.store(userId, args.action, date);
```
Comment (rafal-ksiazek-rmtcfm-com) [inline server/src/services/event/EventService.js:84]: what's this? why we cannot use previous keys?

File: server/src/services/event/EventService.js [hunk 7]
```diff
@@ -1,27 +1,62 @@
...
+                const actionCount =
+                    action === 'buy'
+                        ? await actionService.countBefore(userId, 'register', date)
+                        : await actionService.countAfter(userId, 'buy', date);
+
+                if (actionCount < 1) {
+                    return;
+                }
+
+                return redisService.setBit(`${prefix}:cohort:${key}`, userId, 1);
```
Comment (rafal-ksiazek-rmtcfm-com) [inline server/src/services/event/EventService.js:45]: I think this is completely wrong calculate here in this method.
We should just check during adding action "buy" if user registered before (count bitmap key: {prefix}-bitmap-register-anytime). 
If yes - then add to cohort for this day, week of month, month, year
{prefix}-cohort-set-{timespan}

File: server/src/services/event/EventService.js [hunk 8]
```diff
@@ -1,27 +1,62 @@
...
+            },
+
+            async function (prefix, key, userId, { action }) {
+                if (!action || action !== 'buy' || !key.startsWith('global')) {
+                    return;
+                }
+
+                const actionCount = await actionService.count(userId, 'buy');
```
Comment (rafal-ksiazek-rmtcfm-com) [inline server/src/services/event/EventService.js:53]: The same here,
we need just count bitmap {prefix}-bitmap-buy-{timespan today}
if 0 then
count bitmap {prefix}-bitmap-buy-anytime
if 1 then
add to retention list {prefix}-list-retention:buy-{timestamp}

File: server/src/services/event/EventService.js (hunk 6 above)
Comment (rafal-ksiazek-rmtcfm-com) [reply server/src/services/event/EventService.js:84]: why the hell we have another service for storing things :/ this has no sense!

**Comment by adrian-larysz-rmtcfm-com** [comment]
> why this method is named: generatedArrayFromBits, but inside you are using SETBIT
> 
> so it's not only generatedArray but also changing the bitmap?

Yes. It changes bitmap. But this method is deprecated. I just copied this function in another place in file. But if you want to know, you can specify offset of BITPOS redis command, but this offset is counted by bytes, not bits. So by setting offset to 1, you are checking bit from 8 or 9 bits and skipping previous 7 or 8 bits. So it wouldn't work as expected. But I found a solution. If you found first bit with 1, then store its index in array, then you set it to 0 and then you start searching from the beginning until you get -1, you retrieve indexes of all bits with 1 from bitmap.

**Comment by adrian-larysz-rmtcfm-com** [comment]
> I thought about just "store" method which just store one thing
> 
> ie. store(type, name, userId, timeSpans)
> 
> type: (bitmap, count, set)
> name: (part of the key)
> 
> and then:
> EventService.store(EventService.SET, 'retention:buy', userId, timeSpanService.all(date))
> 
> or
> 
> EventService.store(EventService.BITMAP, 'cohort:buy', userId, [timeSpanService.month(date)])

Gonna try to do it.

**Review by rafal-ksiazek-rmtcfm-com — APPROVED** [review]

**Pull request merged**